clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -244,6 +269,24 @@ synchronized void registerMetric(KafkaMetric metric) {
         return this.metrics;
     }
 
+    /**
+     * This iterates over every Sensor and triggers a removeSensor if it has expired
+     * Package private for testing
+     */
+    class ExpireSensorTask implements Runnable {
+        public void run() {
+            for (Map.Entry<String, Sensor> sensor : sensors.entrySet()) {
+                // removeSensor also locks the sensor object. This is fine because synchronized is reentrant
+                synchronized (sensor.getValue()) {
+                    if (sensor.getValue().isExpired()) {
+                        log.debug("Removing expired sensor {}", sensor.getKey());
+                        removeSensor(sensor.getKey());
```

Comment:
Minor: race condition between record and recreate? This is non-critical, but it seems possible to get an update just before removing the sensor - I haven’t looked closely enough though. That said, if it is indeed an issue but too tricky to address I wouldn’t bother with it.